engine/src/main/java/io/camunda/zeebe/engine/Engine.java

```diff
@@ -130,9 +132,13 @@ public ProcessingResult process(
         return processingResultBuilder.build();
       }
 
-      final boolean isNotOnBlacklist =
-          !processingState.getBlackListState().isOnBlacklist(typedCommand);
-      if (isNotOnBlacklist) {
+      // There is no blacklist check needed if the intent is not instance related
+      // nor if the intent is to create new instances, which can't be blacklisted yet
+      final boolean noBlacklistCheckNeeded =
+          !(record.getIntent() instanceof ProcessInstanceRelatedIntent)
+              || record.getIntent() instanceof ProcessInstanceCreationIntent;
+      if (noBlacklistCheckNeeded
+          || !processingState.getBlackListState().isOnBlacklist(typedCommand)) {
```

Comment:
:wrench: Wouldn't the blacklist check be executed either way because we're using an OR here instead of AND?

Another more important thing. The logic on lines 137-139 is already implemented in the `DbBlacklistState#shouldBeBlacklisted` method (we call `shouldBlacklistInstanceOnError` there, and for `ProcessInstanceCreationIntent` it will return `false`). So we can use `DbBlacklistState#shouldBeBlacklisted` here.

It might also be better to just move these conditions inside the `DbBlacklistState#isOnBlacklist` method. That way, we can combine the "empty check" with a check if this record could have been blacklisted at all. Hope this makes sense.

Reply:
I just read [THIS](https://github.com/camunda/zeebe/pull/12306#discussion_r1165737171), so feel free to continue the related discussion in that thread.

I still think we should re-use the `DbBlacklistState#shouldBeBlacklisted` method here. Otherwise, we might forget to update this code in the future, if we decide to change the conditions for blacklisting a process instance.

Reply:
`shouldBeBlacklisted` has a different semantic, it's about whether a record can cause an instance to get blacklisted, the logic we need here is `do we need to check for blacklisted entries for this record at all`. E.g. `ProcessInstanceCreationIntent#shouldBlacklistInstanceOnError` would return true here which is not what we need.
Thus we can't reuse it 1:1 here.

Reply:
:thinking: I see your point. Maybe we can rename that method so that it's usable in both cases?

My reason is that the `shouldBeBlacklisted` name is also not semantically correct. The method checks if a record for a given intent is "blacklistable" or not. So it doesn't fit with the name. And what the method does fits both use cases.

Reply:
There is a general issue to reevaluate/revisit the whole blacklist concept (maybe starting be reconsidering the naming) and implementation https://github.com/camunda/zeebe/issues/5121 .
@Zelldon & me discussed we would not invest time now making it better given the whole concept is to be reconsidered.

Thus the limited scope with rather primitive measures to avoid it's potential impact on the performance.

Would that be acceptable in this light from that perspective?

Reply:
I think would make sense to have the check of being blacklistable or something in a separate method to reuse it in both places. But I don't see it as a blocker.